File: herencia-dto/SergioDeIscar-herencia-java/SergioDeIscar-herencia-java/src/main/java/org/example/repositories/persona/PersonaRepositoryMap.java
package org.example.repositories.persona;

import org.example.models.Persona;
import org.example.services.storage.persona.PersonaStorageService;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class PersonaRepositoryMap implements PersonaRepository{
    private final Map<Integer, Persona> personas = new LinkedHashMap<>();
    private final PersonaStorageService storageService;

    public PersonaRepositoryMap(PersonaStorageService storageService) {
        this.storageService = storageService;
    }

    @Override
    public List<Persona> getAll() {
        return personas.values().stream().toList();
    }

    @Override
    public Persona getById(Integer integer) {
        return personas.get(integer);
    }

    @Override
    public Persona save(Persona element, boolean storage) {
        personas.put(element.getId(), element);
        if (storage) downgrade();
        return element;
    }

    @Override
    public void saveAll(List<Persona> elements, boolean storage) {
        elements.forEach(e -> save(e, storage));
    }

    @Override
    public Persona deleteById(Integer integer) {
        return personas.remove(integer);
    }

    @Override
    public Persona update(Persona element) {
        return updateById(element.getId(), element);
    }

    @Override
    public Persona updateById(Integer integer, Persona element) {
        if (personas.get(integer) == null) return null;
        personas.put(integer, element);
        return element;
    }

    @Override
    public List<Persona> upgrade() {
        personas.clear();
        var load = storageService.loadAll();
        saveAll(load, false);
        return load;
    }

    @Override
    public List<Persona> downgrade() {
        return storageService.saveAll(getAll());
    }
}

File: herencia-dto/SergioDeIscar-herencia-kotlin/SergioDeIscar-herencia-kotlin/src/main/kotlin/repositories/persona/PersonaRepositoryMap.kt
package repositories.persona

import models.Alumno
import models.Persona
import models.Profesor
import mu.KotlinLogging
import service.storage.persona.PersonaStorageService

private val logger = KotlinLogging.logger {  }

class PersonaRepositoryMap(
    private val storageService: PersonaStorageService
): PersonaRepository {
    private val personas = mutableMapOf<Int, Persona>()
    override fun getPorcentajePorTipo(): Map<String, Double> {
        logger.debug { "PersonaRepositoryMap ->\tgetPorcentajePorTipo" }
        return personas.values
            .groupBy {
                when(it){
                    is Alumno -> "Alumno"
                    is Profesor -> "Profesor"
                    else -> throw RuntimeException("Tipo de persona no reconocido")
                }
            }
            .mapValues { it.value.size.toDouble() / personas.size }
    }

    override fun getAll(): List<Persona> {
        logger.debug { "PersonaRepositoryMap ->\tgetAll" }
        return personas.values.toList()
    }

    override fun getById(id: Int): Persona? {
        logger.debug { "PersonaRepositoryMap ->\tgetById" }
        return personas[id]
    }

    override fun save(element: Persona, storage: Boolean): Persona {
        logger.debug { "PersonaRepositoryMap ->\tsave" }
        personas[element.id] = element
        if (storage) downgrade()
        return element
    }

    override fun saveAll(elements: List<Persona>, storage: Boolean) {
        logger.debug { "PersonaRepositoryMap ->\tsaveAll" }
        elements.forEach { save(it, false) }
        if (storage) downgrade()
    }

    override fun deleteById(id: Int): Persona? {
        logger.debug { "PersonaRepositoryMap ->\tdeleteById" }
        return personas.remove(id)
    }

    override fun update(element: Persona): Persona? {
        logger.debug { "PersonaRepositoryMap ->\tupdate" }
        return updateById(element.id, element)
    }

    override fun updateById(id: Int, element: Persona): Persona? {
        logger.debug { "PersonaRepositoryMap ->\tupdateById" }
        personas[id] ?: return null
        personas[id] = element
        return element
    }

    override fun upgrade(): List<Persona> {
        logger.debug { "PersonaRepositoryMap ->\tupgrade" }
        personas.clear()
        val load = storageService.loadAll()
        saveAll(load, false)
        return load
    }

    override fun downgrade(): List<Persona> {
        logger.debug { "PersonaRepositoryMap ->\tdowngrade" }
        return storageService.saveAll(personas.values.toList())
    }
}